### PR TITLE
Implement custom timeout value feature

### LINKED ISSUE #7: Custom timeout values
## Summary

Currently `tcping` is having a hard-coded timeout value of 1 second. Being able to change this value through a flag can be useful for some scenarios but requires a major refactoring.

### PR DESCRIPTION
**Pull request by luca-patrignani**
# Important

**Please provide the requested parameters below to help us review your pull request.**

**Pull requests without a body or explanation will be rejected.**

## Describe your changes
Implement custom dial timeout feature
## Issue ticket number and link
#7
Closes #7

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added tests.
- [ ] I have run `make check` and there are no failures.

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by pouriyajamshidi** [comment]
Hi @luca-patrignani,

Thanks a lot for your contribution.

It seems like the CodeFactor does not like this line:

image


I will start reviewing this PR once CodeFactor is happy :)

**Comment by pouriyajamshidi** [comment]
@luca-patrignani, I ran a few tests on my machine and noticed some inaccuracies in the time report. For instance running `go run ./... yahoo.com 443 -t 2` yields what is depicted in the image below:

image


we can see `longest consecutive downtime` is wrong. It should be 10. 

Additionally, this part:

```console
TCPing started at: 2023-08-14 16:49:08
TCPing ended at:   2023-08-14 16:49:13
duration (HH:MM:SS): 00:00:10
```

shows the correct duration but `start` and `end` time are not adding up.


I thought this could be something that was already a problem in the code, but after a few checks on v2, I could confirm that the time display is working fine there.

**Comment by pouriyajamshidi** [comment]
> Hi, I'm not able to reproduce the bug, can you help me? And sorry for this long PR review.

Hi @luca-patrignani,

First of all, thank you for willing to contribute to this project and your patience addressing my comments.

what you can run to get that result is:

```bash
go run ./... distrowatch.com 444 -c 4
```

Notice the port `444`. It is intentional to have probe failures to get that result.

There is no rush, take your time.

**Comment by luca-patrignani** [comment]
Hi, I still can't reproduce the bug. This is the output of  `go run ./... distrowatch.com 444 -c 4`:
image
I ran this multiple times but the output always looked correct to me.

**Comment by pouriyajamshidi** [comment]
> Hi, I still can't reproduce the bug. This is the output of `go run ./... distrowatch.com 444 -c 4`: image I ran this multiple times but the output always looked correct to me.

Hi @luca-patrignani,

hmmm, maybe I tested it right before you pushed some changes.

Will test it again tomorrow when I am back.

Thanks again

File: tcping.go [hunk 1]
```diff
@@ -637,24 +649,26 @@ func tcping(tcpStats *stats) {
...
+	conn, err := net.DialTimeout("tcp", IPAndPort.String(), tcpStats.userInput.timeout)
+	connDuration := time.Since(connStart)
+	rtt := nanoToMillisecond(connDuration.Nanoseconds())
 
 	if err != nil {
-		tcpStats.handleConnError(connStart)
+		tcpStats.handleConnError(connStart, maxDuration(connDuration, tcpStats.intervalBetweenProbes))
 	} else {
-		tcpStats.handleConnSuccess(rtt, connStart)
+		tcpStats.handleConnSuccess(rtt, connStart, maxDuration(connDuration, tcpStats.intervalBetweenProbes))
```
Comment (pouriyajamshidi) [inline tcping.go:660]: Let's extract this `maxDuration` outside this condition.

```go
elapsed :=  maxDuration(connDuration, tcpStats.intervalBetweenProbes)
```

Then we just pass it to our conditions

File: tcping.go [hunk 2]
```diff
@@ -578,16 +585,16 @@ func nanoToMillisecond(nano int64) float32 {
 }
 
 // handleConnError processes failed probes
-func (tcpStats *stats) handleConnError(connTime time.Time) {
+func (tcpStats *stats) handleConnError(connTime time.Time, elapsed time.Duration) {
 	if !tcpStats.wasDown {
 		tcpStats.startOfDowntime = connTime
-		calcLongestUptime(tcpStats,
-			time.Duration(tcpStats.ongoingSuccessfulProbes)*time.Second)
+		uptime := time.Since(tcpStats.startOfUptime).Truncate(time.Second)
```
Comment (pouriyajamshidi) [inline tcping.go:591]: I think `Truncate` is causing some very tiny "miscalculations".

Check out the `longest consecutive uptime` value in these two images:

image


And then:


image


You have to run it a few times to reproduce but my guess is, if the connection duration falls under a certain value, it just truncates it.

**Review by pouriyajamshidi — CHANGES_REQUESTED** [review]
Hi @luca-patrignani,

Sorry for my late responses. There are a lot going on simultaneously :)

Overall your work looks good and quite clever.

I have added my remarks which are mostly concerning the time report.

In addition to those, there is one more detail that jumped out:

image

To reproduce, you can use:

```bash
sudo iptables -A OUTPUT -d 74.6.231.21 -j DROP; sleep 3; sudo iptables -F
```

Oddly, instead of moving on to the next probe, it shows 3 seconds in the one that has failed.

**Comment by luca-patrignani** [comment]
> Hi @luca-patrignani,
> 
> Sorry for my late responses. There are a lot going on simultaneously :)
> 
> Overall your work looks good and quite clever.
> 
> I have added my remarks which are mostly concerning the time report.
> 
> In addition to those, there is one more detail that jumped out:
> 
> image
> 
> To reproduce, you can use:
> 
> ```shell
> sudo iptables -A OUTPUT -d 74.6.231.21 -j DROP; sleep 3; sudo iptables -F
> ```
> 
> Oddly, instead of moving on to the next probe, it shows 3 seconds in the one that has failed.

Hi, 
I am not entirely sure of what you would expect. The line should be "No response received for **2** seconds", right?

**Comment by luca-patrignani** [comment]
Hi, 
I just pushed 4 commits, one for each points you made.

**Review by pouriyajamshidi — CHANGES_REQUESTED** [review]
Hey @luca-patrignani,

Everything looks good and this can get merged in when you address the final points.

**Review by pouriyajamshidi — APPROVED** [review]
Good job @luca-patrignani and thanks for your contribution!

Sorry for the extra scrutiny. Anything that touches time is subject to it.

**Pull request merged**